crossplane/README.md

```diff
@@ -0,0 +1,146 @@
+The Secret should contain AWS Access Key.
+
+```yaml
+apiVersion: aws.crossplane.io/v1beta1
+kind: ProviderConfig
+metadata:
+  name: default
+spec:
+  credentials:
+    source: Secret
+    secretRef:
+      namespace: crossplane-system
+      name: aws-creds
+      key: creds
+```
+
+## Design proposals
+
+### 1. Installation
+
+#### Install Crossplane as part of the App collection
+
+This is useful if we want to install Crossplane App on every Management Cluster
+by default. The configuration will live in `giantswarm/config`. This still
+allows us to change available settings and Providers per Cloud Provider and per
+Management Cluster.
+
+#### Install Crossplane as HelmRelease in management-clusters-fleet
+
+This is useful if we want to limit Crossplane to be deployed on select
+Management Clusters. Creating a single `HelmRelease` resource allows us to
+easily set chart values.
+
+### 2. Handling configuration and secrets
+
+#### Giant Swarm stores customer secrets
+
+In this case, Giant Swarm is responsible for all three stages of deployment,
+including Provider Configuration. This means we need to store sensitive data
+(AWS Access Key, GCP Account Keyfile, Azure Service Principal Keys, etc.),
+support their encryption and decryption, and deliver them to Management
+Clusters.
+
+#### Giant Swarm manages Crossplane, the customer provides the keys
+
+In this case, Giant Swarm is responsible for stages one and two. Our goal is to
+deliver Crossplane and a set of Providers. It is up to the customer to create a
+`Secret` and `ProviderConfig`. The structure of both resources is simple and
+well-documented. It also makes sense given how the `ProviderConfig` definition
+varies, depending on the Provider.
+
+### 3. Providers
+
+#### The list of Providers is pre-defined and immutable
+
+We deploy a provider that best matches the Cloud Provider's infrastructure:
+- AWS <https://github.com/crossplane-contrib/provider-aws>
+- Azure <https://github.com/crossplane-contrib/provider-azure>
+- CAPA <https://github.com/crossplane-contrib/provider-aws>?
+- GCP <https://github.com/crossplane/provider-gc>
+- KVM is not supported; obviously, it's the customers' infrastructure
+- OpenStack TBD (<https://github.com/crossplane-contrib/provider-terraform>???)
+
+No other providers are available, even upon request. The offering is "managed".
+
+#### Customers can pick and choose from a well-defined list of vetted Providers
+
+We offer a curated selection of [all Crossplane
+providers](https://github.com/orgs/crossplane-contrib/repositories). Clients
+are free to pick the ones they like for us to install. The offering is "managed".
+
+#### Customers can request any Provider they like
+
+Customers can select any of the Providers they like to have. It means
+Giant Swarm tests and vets the Provider to be able to support the customer and
+"manage" the deployment.
+
+#### Customers can install any Provider they like
```

Comment:
To do this we only need to give access to a [handful of resources](https://github.com/crossplane/crossplane/tree/master/cluster/crds), which we need to do anyway I think.

This gives a lot of space and freedom for the customer I think and makes it easier for us to manage Crossplane. With managing the providers too, even a fixed subset - which constrains the customer - we need to manage a map of provider configs specific to the given providers. I think this does not mean managing the secrets, but they basically do not work without the secrets. I don't think GS should manage the secrets so it makes no sense and just extra work - so to say - to mange the provider configs as they basically add value only that a single ~20 line long yaml is not needed to be managed by the customer. And worst on top of that is that it will bring the same mistake we committed with e.g. `flux-app` that we manage the app, the CRDs and the CRs using those CRDs in the same Helm Chart.

So in short I feel like just managing crossplane is the best way and give customers access to those CRDs, their cluster admins can install whatever provider they need and manage the 2 simple configs (provider and secret) for it. (Actually maybe maybe they want to manage more, cos it could be useful when you want to give different resources different permissions, I think this is one great benefit of the creds / perms being in. the provider config in Crossplane, decoupled and you just tell which one to use when you want to use it). Customers having access to the provider logs they can debug issues too.

The only drawback I see here is that since whatever provider can be installed, the usual security risks and the stability of the MC because of this. But since we want to give more access / control over the MC to customers it may not be an issue (or issue for another day).

Reply:
We can't allow customers to install any provider right now, at least not without some security tuning, that we might not have time for right now. If we allow the customers to install any provider, it's enough to install `provider-kubernetes` to extract any object in the cluster. So, to get started and complete our MVP on time, I suggest to limit the list of  providers and block installation by customers with kyverno